File: decorator_pattern/decorator.cpp
#include<iostream>
#include<string>
using namespace std;

// Component interface
class Coffee{
public:
    virtual string getDescrition() = 0;
    virtual double getCost() = 0;
};


// Concrete Component : Implements the Component interface
class PlainCoffee : public Coffee{
public:
   
    string getDescrition() override {
        return "This is Plain Coffee";
    }
    double getCost() override {
        return 2.0;
    }
};

// Decorator interface
class CoffeeDecorator :public Coffee{
protected:
    Coffee * decoratedCoffee;
public:
    CoffeeDecorator(Coffee * coffee){
        decoratedCoffee = coffee;
    }
    virtual string getDescrition() override{
            return decoratedCoffee->getDescrition();
        }
        virtual double getCost() override{
            return decoratedCoffee->getCost();
        }
};

// decorator implementation
class SugarDecorator : public CoffeeDecorator{
public:
    SugarDecorator(Coffee* coffee) : CoffeeDecorator(coffee) {} // calling the super constructor

    string getDescrition() override {
        return CoffeeDecorator::getDescrition() + ", Milk";
    }
    double getCost() override {
        return CoffeeDecorator::getCost() + 0.5;
    }
};

class Whipped : public CoffeeDecorator{
    public:
    Whipped(Coffee * coffee) : CoffeeDecorator(coffee){}
    string getDescrition() override{
        return CoffeeDecorator::getDescrition() + ", Capachino";
    }
    double getCost() override{
        return CoffeeDecorator::getCost() + 1.0;
    }
};

int main(){
    // get whipped + sugar

    // 1. Start with base coffee
    Coffee *myCoffee = new PlainCoffee();
    // 2. Add Sugar
    myCoffee = new SugarDecorator(myCoffee);
    // 3. Add Capachino
    myCoffee = new Whipped(myCoffee);

    cout << "Order: " << myCoffee->getDescrition() << endl;
    cout << "Total Cost: $" << myCoffee->getCost() << endl;

    // Cleanup (In real HLD/LLD, use smart pointers like std::unique_ptr!)
    delete myCoffee; 
    return 0;
}

File: decorator_pattern/decorator.java
interface Coffee{
    String getDescription();
    double getCost();
};

class PlainCoffee implements Coffee{
    @Override
    public String getDescription(){
        return "Plain Coffee";
    }
    @Override
    public double getCost(){
        return 1.0;
    }
};

abstract class CoffeeDecorator implements Coffee{
    protected Coffee coffee;
    public CoffeeDecorator(Coffee coffee){
        this.coffee = coffee;
    }
    @Override
    public String getDescription(){
        return coffee.getDescription();
    }
    @Override 
    public double getCost(){
        return coffee.getCost();
    }
};

class WhippedCream extends CoffeeDecorator {
    public WhippedCream(Coffee coffee) {
        super(coffee); // Calls the CoffeeDecorator constructor
    }

    @Override
    public String getDescription() {
        return super.getDescription() + ", Whipped Cream";
    }

    @Override
    public double getCost() {
        return super.getCost() + 0.75;
    }
}

// 5. Another Decorator for variety
class Milk extends CoffeeDecorator {
    public Milk(Coffee coffee) {
        super(coffee);
    }

    @Override
    public String getDescription() {
        return super.getDescription() + ", Milk";
    }

    @Override
    public double getCost() {
        return super.getCost() + 0.50;
    }
}



class decorator{
    public static void main(String []args){
        // Create a plain coffee
        Coffee myOrder = new PlainCoffee();
        
        // Wrap it in Milk
        myOrder = new Milk(myOrder);
        
        // Wrap it in Whipped Cream
        myOrder = new WhippedCream(myOrder);

        System.out.println("Order: " + myOrder.getDescription());
        System.out.println("Total Cost: $" + myOrder.getCost());
    }
}